### PR TITLE
Added JWT "CA" and API to Auth Server.

### PR DESCRIPTION
**Pull request by russjones**
**Description**

Added a JWT "CA" to the Auth Server. The JWT CA is not a real CA in that it does not sign certificates, but it behaves like one in terms of storage, access, and rotation.

Added `GenerateJWT` GRPC API to the Auth Server. Used by AAP web server to request signed JWT tokens which will be embedded within the request headers and sent to the target application.

Added `/webapi/certs` endpoint to proxy web server to allow fetching of JWT public keys which can be used to verify the JWT token by the target application.

**Related Issues**

https://github.com/gravitational/teleport/issues/3951

### CODE REVIEW

File: lib/auth/tls_test.go [hunk 1]
```diff
@@ -1766,6 +1897,68 @@ func (s *TLSSuite) TestGenerateCerts(c *check.C) {
...
+			outError:      true,
+		},
+		{
+			inMachineRole: teleport.RoleWeb,
+			inReqUsername: "foo@example.com",
+			inReqRoles:    []string{"bar", "baz"},
+			outError:      false,
+		},
+	}
+	for _, tt := range tests {
```
Comment (awly) [inline lib/auth/tls_test.go:1940]: failures need to somehow distinguish between test cases

File: lib/services/authority.go [hunk 2]
```diff
@@ -408,6 +441,32 @@ func (c *CertAuthorityV2) GetTLSKeyPairs() []TLSKeyPair {
...
+	key, err := jwt.New(&jwt.Config{
+		PublicKey:  c.Spec.JWTKeyPairs[0].PublicKey,
+		PrivateKey: c.Spec.JWTKeyPairs[0].PrivateKey,
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return key, nil
+}
```
Comment (awly) [inline lib/services/authority.go:458]: how will the proxy verify both old and new tokens during rotation?
maybe we need an aggregate version of `jwt.Key`, which signs with one key but verifies with multiple?

File: lib/services/authority.go [hunk 3]
```diff
@@ -616,10 +675,34 @@ func (ca *CertAuthorityV2) SetSigningAlg(alg string) {
...
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+func (ca *CertAuthorityV2) checkUserOrHostCA() error {
 	err := ca.ID().Check()
 	if err != nil {
 		return trace.Wrap(err)
 	}
```
Comment (awly) [inline lib/services/authority.go:697]: move this to the `Check` method above, it's the same for all CAs

File: lib/services/authority.go [hunk 4]
```diff
@@ -638,6 +722,42 @@ func (ca *CertAuthorityV2) Check() error {
 	return nil
 }
 
+func (ca *CertAuthorityV2) checkJWT() error {
```
Comment (awly) [inline lib/services/authority.go:725]: nit: `checkJWTKeys`

File: lib/services/suite/suite.go [hunk 5]
```diff
@@ -85,6 +91,12 @@ func NewTestCA(caType services.CertAuthType, clusterName string, privateKeys ...
 			CheckingKeys: [][]byte{ssh.MarshalAuthorizedKey(signer.PublicKey())},
 			SigningKeys:  [][]byte{keyBytes},
 			TLSKeyPairs:  []services.TLSKeyPair{{Cert: cert, Key: key}},
+			JWTKeyPairs: []services.JWTKeyPair{
+				services.JWTKeyPair{
```
Comment (awly) [inline lib/services/suite/suite.go:95]:
```suggestion
				{
```

File: lib/services/trust.go [hunk 6]
```diff
@@ -72,14 +72,19 @@ const (
 	HostCA CertAuthType = "host"
 	// UserCA identifies the key as a user certificate authority
 	UserCA CertAuthType = "user"
+	// JWT identifies type of certificate authority as JWT. In this case JWT is
+	// not a real certificate authority because it does not issue certificates
+	// but rather an entity that can be used to issue JWT tokens, however it
+	// behaves much like a CA in terms of rotation and storage.
+	JWT CertAuthType = "jwt"
```
Comment (awly) [inline lib/services/trust.go:79]: how about `JWTSigner`?

File: lib/services/types.proto [hunk 7]
```diff
@@ -185,6 +185,14 @@ message TLSKeyPair {
     bytes Key = 2 [ (gogoproto.jsontag) = "key,omitempty" ];
 }
 
+// JWTKeyPair is a PEM encoded keypair used for signing JWT tokens.
+message JWTKeyPair {
+    // PublicKey is a PEM encoded public key.
+    bytes PublicKey = 1 [ (gogoproto.jsontag) = "public_key,omitempty" ];
+    // PrivateKey is a PEM encoded private key. Also contains public key.
```
Comment (awly) [inline lib/services/types.proto:192]: does it contain public key directly?
you can derive it, but the public key isn't directly present as a second PEM blob

File: lib/services/types.proto [hunk 8]
```diff
@@ -238,6 +246,9 @@ message CertAuthoritySpecV2 {
         RSA_SHA2_512 = 3;
     }
     SigningAlgType SigningAlg = 9 [ (gogoproto.jsontag) = "signing_alg,omitempty" ];
+    // JWTKeyPair is a list of JWT key pairs.
+    repeated JWTKeyPair JWTKeyPairs = 10
+        [ (gogoproto.nullable) = false, (gogoproto.jsontag) = "jwt_key_pairs,omitempty" ];
```
Comment (awly) [inline lib/services/types.proto:251]: why can't this be nullable?

File: lib/web/apiserver.go [hunk 9]
```diff
@@ -683,6 +686,30 @@ func (h *Handler) getWebConfig(w http.ResponseWriter, r *http.Request, p httprou
 	return nil, nil
 }
 
+type certsResponse struct {
+	JWTKeys []services.JWTKeyPair `json:"jwt_keys"`
+}
+
+func (h *Handler) certs(w http.ResponseWriter, r *http.Request, p httprouter.Params) (interface{}, error) {
```
Comment (awly) [inline lib/web/apiserver.go:693]: nit: rename to `jwtPublicKeys`, there are no certs here

File: lib/auth/rotate.go [hunk 10]
```diff
@@ -481,116 +495,174 @@ func startNewRotation(req rotationReq, ca services.CertAuthority) error {
...
+	tlsKeyPairs := ca.GetTLSKeyPairs()
+	jwtKeyPairs := ca.GetJWTKeyPairs()
 
-	signingKeys = [][]byte{signingKeys[1], signingKeys[0]}
-	checkingKeys = [][]byte{checkingKeys[1], checkingKeys[0]}
+	// Perform a length check because not every CA has all types of keys.
+	// For example, a JWT services.CertAuthority will not have SSH or TLS keys.
+	// Similarly SSH and TLS services.CertAuthority do not have JWT keys.
+	if len(signingKeys) > 0 {
+		signingKeys = [][]byte{signingKeys[1], signingKeys[0]}
```
[the remark this answers is not in the document]
Comment (russjones) [reply lib/auth/rotate.go:634]: Was having issues with this, will ping you later.

File: lib/auth/tls_test.go [hunk 11]
```diff
@@ -689,6 +689,137 @@ func (s *TLSSuite) TestRollback(c *check.C) {
 	c.Assert(err, check.IsNil)
 }
 
+// TestJWTRotation checks that JWT tokens can be rotated and tokens can or
+// can not be validated at the appropriate phase.
+func (s *TLSSuite) TestJWTRotation(c *check.C) {
```
[the remark this answers is not in the document]
Comment (russjones) [reply lib/auth/tls_test.go]: Will do.

File: lib/auth/proto/auth.proto [hunk 12]
```diff
@@ -235,6 +235,26 @@ message DeleteAllAppsRequest {
...
+    // Namespace is the namespace for the registered application.
+    string Namespace = 1 [ (gogoproto.jsontag) = "namespace" ];
+    // Username is the Teleport identity of the user for who a JWT is being
+    // issued.
+    string Username = 2 [ (gogoproto.jsontag) = "username" ];
+    // Roles is the list of roles for the Teleport identity for whom a JWT
+    // is being issued.
+    repeated string Roles = 3 [ (gogoproto.jsontag) = "roles" ];
+    // Recipient is the target.
+    string Recipient = 4 [ (gogoproto.jsontag) = "recipient" ];
```
Comment (awly) [inline lib/auth/proto/auth.proto:250]: Is this the application name?
If yes, update the field name and/or comment.

File: lib/auth/rotate.go [hunk 13]
```diff
@@ -454,9 +459,12 @@ func processRotationRequest(req rotationReq) (services.CertAuthority, error) {
 	}
 }
 
-// startNewRotation starts new rotation and updates the certificate
-// authority with new CA keys.
+// startNewRotation starts new rotation. In this phase requests will continue
+// to be signed by the old CA (index 0), but a new CA (index 1) will be added.
+// This new CA can be used to verify requests.
 func startNewRotation(req rotationReq, ca services.CertAuthority) error {
+	var err error
```
Comment (awly) [inline lib/auth/rotate.go:466]: define the error in the smallest possible scope, to help with shadowing bugs

File: lib/auth/rotate.go [hunk 14]
```diff
@@ -481,116 +496,174 @@ func startNewRotation(req rotationReq, ca services.CertAuthority) error {
...
 			return trace.Wrap(err)
 		}
+
+		jwtPublicKey, jwtPrivateKey, err = jwt.GenerateKeyPair()
+		if err != nil {
+			return trace.Wrap(err)
+		}
 	}
 
 	tlsKeyPair := &services.TLSKeyPair{
```
Comment (awly) [inline lib/auth/rotate.go:534]: nit: make this a non-pointer, to match `jwtKeyPair`

File: lib/auth/rotate.go [hunk 15]
```diff
@@ -481,116 +496,174 @@ func startNewRotation(req rotationReq, ca services.CertAuthority) error {
...
+		if len(signingKeys) > 0 {
+			signingKeys = [][]byte{signingKeys[0], sshPrivateKey}
+			checkingKeys = [][]byte{checkingKeys[0], sshPublicKey}
+		}
+		if len(tlsKeyPairs) > 0 {
+			tlsKeyPairs = []services.TLSKeyPair{tlsKeyPairs[0], *tlsKeyPair}
+		}
+		if len(jwtKeyPairs) > 0 {
+			jwtKeyPairs = []services.JWTKeyPair{jwtKeyPairs[0], jwtKeyPair}
+		}
```
Comment (awly) [inline lib/auth/rotate.go:590]: use `append` here instead

File: lib/auth/rotate.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (awly) [reply lib/auth/rotate.go:634]: what was the issue?

File: lib/auth/rotate.go [hunk 16]
```diff
@@ -599,40 +672,62 @@ func startRollingBackRotation(ca services.CertAuthority) error {
...
 
-	// Rotation sets the first key to be the new key
-	// and keep only public keys/certs for the new CA.
-	signingKeys = [][]byte{signingKeys[1]}
-	checkingKeys = [][]byte{checkingKeys[1]}
+	// Perform a length check because not every CA has all types of keys.
+	// For example, a JWT services.CertAuthority will not have SSH or TLS keys.
+	// Similarly SSH and TLS services.CertAuthority do not have JWT keys.
+	if len(signingKeys) > 0 {
+		signingKeys = [][]byte{signingKeys[1]}
```
Comment (awly) [inline lib/auth/rotate.go:682]:
```suggestion
		signingKeys[0], signingKeys[1] = signingKeys[1], signingKeys[0]
```

File: lib/auth/rotate.go [hunk 17]
```diff
@@ -481,116 +496,174 @@ func startNewRotation(req rotationReq, ca services.CertAuthority) error {
...
-	keyPairs := ca.GetTLSKeyPairs()
+	tlsKeyPairs := ca.GetTLSKeyPairs()
+	jwtKeyPairs := ca.GetJWTKeyPairs()
 
-	signingKeys = [][]byte{signingKeys[1], signingKeys[0]}
-	checkingKeys = [][]byte{checkingKeys[1], checkingKeys[0]}
+	// Perform a length check because not every CA has all types of keys.
+	// For example, a JWT services.CertAuthority will not have SSH or TLS keys.
+	// Similarly SSH and TLS services.CertAuthority do not have JWT keys.
+	if len(signingKeys) > 0 {
```
Comment (awly) [inline lib/auth/rotate.go:633]: Should the checks for swapping keys be `if len(keys) > 1`?
There should be at least 2 keys to swap them without a panic.

File: lib/auth/tls_test.go [hunk 18]
```diff
@@ -2631,3 +2835,38 @@ func (s *TLSSuite) TestAPIBackwardsCompatibilityLogic(c *check.C) {
...
+			Username:  "foo",
+			Recipient: "panel",
+		})
+		if err != nil {
+			errs = append(errs, trace.Wrap(err))
+			continue
+		}
+		if err == nil {
+			return claims, nil
+		}
```
Comment (awly) [inline lib/auth/tls_test.go:2869]:
```suggestion
		return claims, nil
```

File: lib/services/authority.go [hunk 19]
```diff
@@ -408,6 +474,37 @@ func (c *CertAuthorityV2) GetTLSKeyPairs() []TLSKeyPair {
 	return c.Spec.TLSKeyPairs
 }
 
+// JWTSigner returns the active JWT key used to sign tokens.
+func (c *CertAuthorityV2) JWTSigner() (*jwt.Key, error) {
+	if len(c.Spec.JWTKeyPairs) == 0 {
+		return nil, trace.BadParameter("not JWT keypairs found")
```
Comment (awly) [inline lib/services/authority.go:480]:
```suggestion
		return nil, trace.BadParameter("no JWT keypairs found")
```

File: lib/web/apiserver.go [hunk 20]
```diff
@@ -157,6 +157,9 @@ func NewHandler(cfg Config, opts ...HandlerOption) (*RewritingHandler, error) {
 	// OIDC connectors and auth preferences
 	h.GET("/webapi/find", httplib.MakeHandler(h.find))
 
+	// Unauthenticated access to public keys.
+	h.GET("/webapi/certs", httplib.MakeHandler(h.jwtPublicKeys))
```
Comment (awly) [inline lib/web/apiserver.go:161]: `/webapi/certs` seems too general.
how about `/webapi/jwt_public_keys`? or `jwt_verification_keys`?

File: lib/web/apiserver.go [hunk 21]
```diff
@@ -683,6 +686,30 @@ func (h *Handler) getWebConfig(w http.ResponseWriter, r *http.Request, p httprou
...
+	JWTKeys []services.JWTKeyPair `json:"jwt_keys"`
+}
+
+func (h *Handler) jwtPublicKeys(w http.ResponseWriter, r *http.Request, p httprouter.Params) (interface{}, error) {
+	clusterName, err := h.cfg.ProxyClient.GetDomainName()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Fetch the host CA public keys only.
```
Comment (awly) [inline lib/web/apiserver.go:699]:
```suggestion
	// Fetch the JWT public keys only.
```

File: lib/web/apiserver.go [hunk 22]
```diff
@@ -683,6 +686,30 @@ func (h *Handler) getWebConfig(w http.ResponseWriter, r *http.Request, p httprou
...
+		Type:       services.JWTSigner,
+		DomainName: clusterName,
+	}, false)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &jwtPublicKeysResponse{
+		JWTKeys: ca.GetJWTKeyPairs(),
+	}, nil
```
Comment (awly) [inline lib/web/apiserver.go:710]: will the private keys get stripped at some point here?

**Review by awly — APPROVED** [review]
Mostly naming comments, LGTM.
One important question on the proxy API endpoint - want to make sure the private keys don't leak

**Comment by benarent** [comment]
While testing out JWTs, it seem it's best practice to place them on 

`https://YOUR_DOMAIN/.well-known/jwks.json` - https://auth0.com/docs/tokens/json-web-tokens/json-web-key-sets 
AWS: `https://cognito-idp.{region}.amazonaws.com/{userPoolId}/.well-known/jwks.json`

**Pull request closed without merging**